File: Data Structures & Algorithms/search-2d-matrix/submission-3.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        h = len(matrix)
        w = len(matrix[0])
        l, r = 0, h*w-1
        while(l <= r):
            cur = (l + r) // 2
            ch = cur // w
            cw = cur % w
            if matrix[ch][cw] == target:
                return True
            elif matrix[ch][cw] < target:
                l = cur + 1
            else:
                r = cur - 1
        return False



        

File: Data Structures & Algorithms/two-integer-sum/submission-3.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> mem;
        for (int i = 0; i < nums.size(); i++) {
            int t = target - nums[i];
            if (mem.contains(t)) {
                vector<int> answer = {mem[t], i};
                return answer;
            }
            else {
                mem.insert({nums[i], i});
            }
        }
        vector<int> answer = {-1, -1};
        return answer;
    }
};
